src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageActionDescriptorProvider.cs

```diff
@@ -57,9 +58,29 @@ protected IList<PageRouteModel> BuildModel()
                 _routeModelProviders[i].OnProvidersExecuted(context);
             }
 
+            RemoveSupersededFallbackRoutes(context);
+
             return context.RouteModels;
         }
 
+        private void RemoveSupersededFallbackRoutes(PageRouteModelProviderContext context)
+        {
+            for (var i = 0; i < context.RouteModels.Count; i++)
+            {
+                var fallbackRouteModel = context.RouteModels[i];
+                if (!fallbackRouteModel.Properties.ContainsKey(FallbackRazorPagesConventions.FallbackPropertyKey))
+                {
+                    continue;
+                }
+
+                if (context.RouteModels.Any(model => FallbackRazorPagesConventions.IsSuperseded(fallbackRouteModel, model)))
```

Comment:
Should we use a dictionary for this? This is O(N^2) doing an O(N) string comparison